ahnlich/ai/src/engine/ai/providers/ort.rs

```diff
@@ -0,0 +1,266 @@
+    }
+}
+
+impl ORTProvider {
+    pub(crate) fn new() -> Self {
+        Self {
+            cache_location: None,
+            cache_location_extension: PathBuf::from("huggingface"),
+            supported_models: None,
+            model: None,
+        }
+    }
+
+    pub fn normalize(v: &[f32]) -> Vec<f32> {
+        let norm = (v.iter().map(|val| val * val).sum::<f32>()).sqrt();
+        let epsilon = 1e-12;
+
+        // We add the super-small epsilon to avoid dividing by zero
+        v.iter().map(|&val| val / (norm + epsilon)).collect()
+    }
+}
+
+impl ProviderTrait for ORTProvider {
+    fn set_cache_location(&mut self, location: &Path) -> &mut Self {
+        self.cache_location = Some(location.join(self.cache_location_extension.clone()));
+        self
+    }
+
+    fn set_model(&mut self, model: &SupportedModels) -> &mut Self {
+        self.supported_models = Some(*model);
+        self
+    }
+
+    fn load_model(&mut self) -> &mut Self {
+        let cache_location = self
+            .cache_location
+            .clone()
+            .expect("Cache location not set.");
+        let supported_model = self.supported_models.expect(
+            &AIProxyError::AIModelNotInitialized.to_string());
+        let ort_model = ORTModel::try_from(&supported_model).unwrap();
+
+        let threads = available_parallelism()
+            .expect("Could not find the threads")
+            .get();
+
+        let cache = Cache::new(cache_location);
+        let api = ApiBuilder::from_cache(cache)
+            .with_progress(true)
+            .build()
+            .unwrap();
+
+        match ort_model {
+            ORTModel::Image(ORTImageModel {
+                weights_file,
+                repo_name,
+                input_param,
+                output_param,
+                ..
+            }) => {
+                let model_repo = api.model(repo_name.clone());
+                let model_file_reference = model_repo
+                    .get(&weights_file)
+                    .unwrap_or_else(|_| panic!("Failed to retrieve {} ", weights_file));
+                let executioners: Vec<ExecutionProviderDispatch> = Default::default();
+                let session = Session::builder()
+                    .unwrap()
+                    .with_execution_providers(executioners)
+                    .unwrap()
+                    .with_optimization_level(GraphOptimizationLevel::Level3)
+                    .unwrap()
+                    .with_intra_threads(threads)
+                    .unwrap()
+                    .commit_from_file(model_file_reference)
+                    .unwrap();
+                self.model = Some(ORTModel::Image(ORTImageModel {
+                    repo_name,
+                    weights_file,
+                    input_param,
+                    output_param,
+                    session: Some(session),
+                }));
+            }
+        }
+        self
+    }
+
+    fn get_model(&self) {
```

Comment:
Same as above.... we would all feel safer if we only unwrap whenever we are absolutely certain it can't break (hence `.expect` or we want it to break the entire flow